### ISSUE TITLE
Importing PostgreSQL Dump Fails with Role Errors

### ISSUE THREAD

**Issue by gitayam** [spantaleev/matrix-docker-ansible-deploy#3684]
I would appreciate any guidance or suggestions on how to resolve these role-related import issues effectively. I am eager to learn if there are recommended steps or commands for preparing the SQL dump for a successful import into PostgreSQL.

This all happened after upgrading Postgres from 16.4 to 17 using the ansible commands from the ./Docs

When running after the upgrade was getting issues like this:

```bash
  delta: '0:00:01.109520'
  end: '2024-10-26 22:43:06.689077'
  failed_when_result: true
  finished: 1
  msg: non-zero return code
  rc: 3
  results_file: /root/.ansible_async/j98691517816.1070525
  start: '2024-10-26 22:43:05.579557'
  started: 1
  stderr: |-
    ERROR:  role "matrix_bot_maubot" does not exist
    STATEMENT:  ALTER DATABASE matrix_bot_maubot OWNER TO matrix_bot_maubot;
  stderr_lines: <omitted>
  stdout: |-
  ```

System Information

- Matrix Server OS: Debian
- Kernel Version: Linux matrix 6.1.0-26-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.112-1 (2024-09-30) x86_64 GNU/Linux

Description

I'm encountering issues while attempting to import a PostgreSQL dump file for my Matrix server setup. Despite several attempts to resolve these issues by modifying the SQL dump file, the import process consistently fails due to roles-related errors.

Steps Tried

1. Commented Out Role Creation and Alteration:
- I used sed to comment out the CREATE ROLE and ALTER ROLE statements in the SQL dump (postgres.sql.gz) to prevent errors related to existing roles. The command I used is as follows:

```bash
sed -i -e 's/CREATE ROLE honoroit;/-- CREATE ROLE honoroit;/g' \
       -e 's/ALTER ROLE honoroit;/-- ALTER ROLE honoroit;/g' \
       -e 's/CREATE ROLE matrix;/-- CREATE ROLE matrix;/g' \
       -e 's/ALTER ROLE matrix;/-- ALTER ROLE matrix;/g' \
       ... # additional role modifications as needed
```

1. Prepared the SQL Dump:
- After commenting out the necessary lines, I compressed the modified SQL dump into a new file (postgres2.sql.gz).
1. Import Attempt via Docker:
- I attempted to import the SQL dump using Docker with the following command:
```bash
docker run --rm --name matrix-postgres-import --log-driver=none --user=1000:1000 --cap-drop=ALL --network=matrix-postgres --env-file=/matrix/postgres/env-postgres-psql --mount type=bind,src=/matrix/postgres.sql.gz,dst=/postgres.sql.gz,ro --entrypoint=/bin/sh docker.io/postgres:17.0-alpine -c "cat /postgres.sql.gz | gunzip | psql -b -v ON_ERROR_STOP=1 -h 172.19.0.2 --dbname=matrix"
```

Errors Encountered

1. Role Does Not Exist Errors:
- Example Error:
```bash
ERROR:  role "matrix_bot_maubot" does not exist
STATEMENT:  ALTER ROLE matrix_bot_maubot WITH NOSUPERUSER INHERIT NOCREATEROLE NOCREATEDB LOGIN NOREPLICATION NOBYPASSRLS PASSWORD '...';
```

- This indicates that the import process is trying to alter a role that has not been created.

1. Duplicate Role Creation Errors:
- Example Error:
```bash
ERROR:  role "matrix" already exists
STATEMENT:  CREATE ROLE matrix;
```

- This error arises because the import process attempts to create a role already in the database.

Additional Attempts

- After facing issues with the previous commands, I commented out both CREATE ROLE and ALTER ROLE statements using the following command, but the import still failed:

```bash
sed -i -e 's/CREATE ROLE honoroit;/-- CREATE ROLE honoroit;/g' \
       -e 's/ALTER ROLE honoroit;/-- ALTER ROLE honoroit;/g' \
       ... # additional role modifications
```

**Comment by gitayam** [comment]
# PostgreSQL Setup and Maintenance Guide for Matrix Server

Solved. 
This guide outlines the steps taken to set up and maintain the PostgreSQL service for the Matrix server, including stopping services, vacuuming the database, and restoring necessary roles and tables.

## Prerequisites
- Ansible installed on the local machine.
- Access to the inventory file for Ansible.
- Necessary roles defined in the `Makefile`.

## Steps

### 1. Stop Synapse Services
To stop the Synapse services, run the following command:

```bash
sudo ansible-playbook -i inventory/hosts setup.yml --tags=stop

2. Remove All PostgreSQL Data

On the Matrix server, remove all PostgreSQL data:

# Assuming the PostgreSQL data directory is at the default location
sudo rm -rf /var/lib/postgresql/data/*

3. Run PostgreSQL Vacuum

After removing the data, run the vacuum command to clean up the database:

sudo ansible-playbook -i inventory/hosts setup.yml --tags=run-postgres-vacuum

4. Set Up PostgreSQL Service

Reconfigure the PostgreSQL service without starting other services:

sudo make roles
sudo ansible-playbook -i inventory/hosts setup.yml --tags=setup-postgres

5. Drop Existing Databases and Types

On the Matrix server, connect to PostgreSQL and drop any existing databases and types:

/matrix/postgres/bin/cli -d postgres

Run the following commands in the PostgreSQL CLI:

DROP DATABASE IF EXISTS synapse;
DROP DATABASE IF EXISTS honoroit;
DROP DATABASE IF EXISTS matrix;
DROP DATABASE IF EXISTS matrix_bot_maubot;
DROP DATABASE IF EXISTS matrix_dimension;
DROP DATABASE IF EXISTS matrix_etherpad;
DROP DATABASE IF EXISTS matrix_mautrix_discord;
DROP DATABASE IF EXISTS matrix_mautrix_signal;
DROP DATABASE IF EXISTS matrix_mautrix_whatsapp;
DROP DATABASE IF EXISTS matrix_prometheus_postgres_exporter;
DROP DATABASE IF EXISTS matrix_registration;
DROP DATABASE IF EXISTS matrix_reminder_bot;
DROP DATABASE IF EXISTS matrix_sliding_sync;
DROP TYPE IF EXISTS public.membership;

6. Restart PostgreSQL Service

Restart the PostgreSQL service to apply changes:

docker restart matrix-postgres

7. Recreate Necessary Databases

Recreate the necessary databases:

/matrix/postgres/bin/cli -d postgres

Run the following commands to create the databases:

CREATE DATABASE synapse;
CREATE DATABASE honoroit;
CREATE DATABASE matrix;
CREATE DATABASE matrix_bot_maubot;
CREATE DATABASE matrix_dimension;
CREATE DATABASE matrix_etherpad;
CREATE DATABASE matrix_mautrix_discord;
CREATE DATABASE matrix_mautrix_signal;
CREATE DATABASE matrix_mautrix_whatsapp;
CREATE DATABASE matrix_prometheus_postgres_exporter;
CREATE DATABASE matrix_registration;
CREATE DATABASE matrix_reminder_bot;
CREATE DATABASE matrix_sliding_sync;

Final Steps

After recreating the databases, you may need to re-import data or roles based on your specific application needs.

**Comment by gitayam** [comment]
Solved . See above

**Comment by spantaleev** [comment]
The playbook does not write any data to `/var/lib/postgresql/data/`, so not sure why your guide clears that.

Postgres as configured by the playbook always writes its data to `/matrix/postgres/data`. When you run `just install-all` or `just setup-all`, Postgres and various databases for all services (and associated users for these databases) are prepared as empty.

------

Importing currently explicitly tries to skip `CREATE DATABASE ..` statements for these same databases, as the playbook knows it has already prepared them and they exist.

If you're seeing an error during importing like:

>  ERROR:  role "matrix_bot_maubot" does not exist
>  STATEMENT:  ALTER DATABASE matrix_bot_maubot OWNER TO matrix_bot_maubot;

.. it's an indication that:

- you have either not executed `install-all` / `setup-all` and are instead immediately trying to import
- you have not enabled the maubot service, so a database hasn't been prepared to it, yet your dump contains definitions for creating/altering and pouring data into this database

-----

Something is off for you. It may be because you have some old databases for services that no longer exist, though.. that still shouldn't break things.

I have done a Postgres upgrade from 16.4 to 17.0 for a few servers already without issues, so I wonder what's different for you.

**Comment by Marx1st** [comment]
Just to add, that we had problems upgrading to PostgreSQL 17, too. We don't have any logs, though as we just rolled back and stay on 16.7 for now.

**Comment by gitayam** [comment]
> you have either not executed install-all / setup-all and are instead immediately trying to import


Yes, this likely would have got things working; following the guide, I went straight to import and didn't stop services and make roles, install, and set up. 

After running unsuccessfully, it was a blank server essentially, and with use, it added data to the DB. 
"
I have updated the path of Postgres deletions. you are right, in searching for it I had come across the path but looking back at my commands you were right
